### PR TITLE
ci: Enable programtests against local backend, improve isolation

### PR DESCRIPTION
**Pull request by AaronFriel**
These changes make integration tests run much more quickly on a local dev loop, for easily parallelized program tests like in the `integration_go_test.go` file the improvement was on the order of minutes across all tests.

The changes rely on setting `PULUMI_BACKEND_URL` to override the backend during particular tests. The backend is set to a temporary directory which is cleaned up on exit. 

A helper function `NewBackendUrl(t *testing.T)` is added to enable

When `PULUMI_TEST_USE_SERVICE=true`, the `RequireService` option is set to true.

When `RequireService == true`, the test is skipped if an access token is not present, improving local dev experience by skipping tests which would error very loudly.

When `RequireService == false and CloudURL == ""`, then we use the helper function to create a temporary directory and point the filestate backend to it. 

The CloudURL check allows tests which, even in the presence of `PULUMI_TEST_USE_SERVICE=true`, to still run against a local backend. E.g.: 

```go
	localTestOptions := testOptions.With(integration.ProgramTestOptions{
		CloudURL: integration.NewBackendURL(t),
	})
```

### CODE REVIEW

File: pkg/testing/integration/program.go
Comment (justinvp) [inline pkg/testing/integration/program.go:210]: Should we also copy this from overrides in the implementation of `With`, e.g.:

https://github.com/pulumi/pulumi/blob/3b843ffbfc8a344881fda0dfcb6881c9be20220a/pkg/testing/integration/program.go#L467-L469

(I think I may have forgotten to do that for some of the fields I've added to `ProgramTestOptions` in the past.)

Comment (AaronFriel) [reply pkg/testing/integration/program.go:210]: Good catch!

**Review by justinvp — APPROVED** [review]
LGTM

File: pkg/testing/integration/program.go
Comment (t0yv0) [inline pkg/testing/integration/program.go:764]: Can we name this better? I look at the name and it gives me no clue this opts me into using a filestate backend in a tempdir.
 
`NewTempFilestateBackendURL`? 

A bit verbose.

File: pkg/testing/integration/program.go
Comment (t0yv0) [inline pkg/testing/integration/program.go:747]: Bits of new magic in here? It will now default to filestate unless the user sets flags? And what about the skip behavior this is new?

File: changelog/pending/20220914--ci--local-programtest.yaml
Comment (t0yv0) [inline changelog/pending/20220914--ci--local-programtest.yaml:4]: Skimming the PR, it feels like this is changing defaults to users of ProgramTest framework - that might be significant and needs to be called out.

**Comment by t0yv0** [comment]
```
When RequireService == false and CloudURL == "", then we use the helper function to create a temporary directory and point the filestate backend to it.
```

This is the default behavior - as most users never set either of those.... and it's changing now to use filestate. This might be massive. For example this change when rolled out will change the behavior of benchmarks I've just been setting up.

File: pkg/testing/integration/program.go
Comment (AaronFriel) [reply pkg/testing/integration/program.go:747]: New magic, yeah. This should make the local dev experience when running `make` smoother, as tests will run against filestate if possible, and skip if not.

In our CI harness, we always set `PULUMI_ACCESS_TOKEN`, so no skipping in CI.

File: changelog/pending/20220914--ci--local-programtest.yaml
Comment (AaronFriel) [reply changelog/pending/20220914--ci--local-programtest.yaml:4]: Good call.

File: pkg/testing/integration/program.go
Comment (AaronFriel) [reply pkg/testing/integration/program.go:764]: `MakeTempBackend`, which is reminiscent of `MkDirTemp`?

File: changelog/pending/20220914--ci--local-programtest.yaml
Comment (AaronFriel) [reply changelog/pending/20220914--ci--local-programtest.yaml:4]: I'll announce to #engineering with a request-for-comment. This does impact other users of ProgramTest ([GitHub code search](https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Apulumi+%2F%5CbProgramTest%5Cb%2F#)) so it's worth getting feedback and figuring out how to default this correctly.

I _believe_ that providers repos, there are several, shouldn't be testing against our service. It introduces more failure modes into what should just be a test of the provider. Of course, that's my opinion and the code owners there should weigh in.

Comment (t0yv0) [reply changelog/pending/20220914--ci--local-programtest.yaml:4]: Yup! Thank you. I'm ok with the new default just it need to be highlighted in the CHANGELOG imo, but also happy if you want to get more feedback first. If there's a way to get this change in without changing the default it's an even easier conversation.

File: pkg/testing/integration/program.go
Comment (t0yv0) [reply pkg/testing/integration/program.go:747]: Ah nice, maybe worth a quick in-source comment.

File: changelog/pending/20220914--ci--local-programtest.yaml
Comment (AaronFriel) [reply changelog/pending/20220914--ci--local-programtest.yaml]: Good endorsement from the providers team on this, and it won't have any immediate impact. If they have tests that require a full service (e.g.: stackreferences), they can address case by case.

**Review by t0yv0 — APPROVED** [review]

**Pull request merged**
